Improve Philips Hue color conversion

starkillerOG opened the pull request:
## Description:
The Philips Hue lights do not support the ordinarry color gammut, theirfore when a color is selected that the philips hue light cannot produce, the closest color that can be produced needs to be found.
Using a rgb to hs conversion that was made specifically for the philips hue lights, much closer colors can be produced.
This is most obvious when selecting pure blue: RGB:[0,0,255].
This PR theirfore improves color selection when selecting collors with high saturation (HSV with S=100%).
colors within the capability of the Hue lights are not changed by this PR, they will still be identical.

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#<home-assistant.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [x] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [x] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [x] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [x] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [x] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [x] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [x] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [x] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard.py#L54

starkillerOG committed:
Improve Philips Hue color conversion

The Philips Hue lights do not support the ordinarry color gammut, theirfore when a color is selected that the philips hue light cannot produce, the closest color that can be produced needs to be found.
Using a rgb to hs conversion that was made specifically for the philips hue lights, much closer colors can be produced.
This is most obvious when selecting pure blue: RGB:[0,0,255].
This PR theirfore improves color selection when selecting collors with high saturation (HSV with S=100%).
colors within the capability of the Hue lights are not changed by this PR, they will still be identical.

starkillerOG committed:
fix white spaces

starkillerOG committed:
fix line to long

starkillerOG committed:
Add rgbxy for Philips Hue

starkillerOG committed:
fix import error

starkillerOG committed:
fix white spaces

starkillerOG committed:
Fix __init__ document string

starkillerOG committed:
fix snake_case naming

starkillerOG committed:
fix line to long

starkillerOG committed:
Add rgbxy to requirements_test for philips hue

starkillerOG committed:
add rgbxy to gen_requirements_all test

amelchio commented on homeassistant/components/light/hue.py:
```diff
@@ -330,7 +336,9 @@ def device_info(self):
-                command['xy'] = color.color_hs_to_xy(*kwargs[ATTR_HS_COLOR])
+                rgb = color.color_hs_to_RGB(*kwargs[ATTR_HS_COLOR])
+                xy_color = self.hue_converter.rgb_to_xy(rgb[0], rgb[1], rgb[2])
+                command['xy'] = xy_color
```
Your issue is probably here since we do not actually set `rgb(0,0,255)`. This is due to #13665.

starkillerOG replied:
Well actually this is not my issue, I tryed using the `hue` and `sat` directly as commands to the hue bridge, but that makes the bulb colors even worse off from what I selected. I am still sending ''xy' color to the bulb just as original, but  the problem I am having lays in the fact that some of my phiplips hue lights can not produce the whole color space that normal "hue" "sat" covers. When selecting a color that lies within the capabilities of the light everything is kind of fine I think and that color is produced by the light (in the current code and with this new PR). 

However when selecting a color that lies outside of the capabilities of the light things go wrong. somehow the closest color needs to be found. Apperently letting the hue bridge decide wich color that schould be gives poor results with a color representing more purple than blue when selecting rgb(0,0,255). When I convert the color using the correct color gamut as specified by the Hue bridge for that light, results are way better and the light actually turns pretty blue for rgb(0,0,255).

amelchio reviewed (changes requested):
This is opening a can of worms, there are so many past issues with Hue colors.

Please give more details about the problem that this is fixing.

starkillerOG committed:
Retrieve color gamut type from Hue bridge

Add code to get the color gamut type for each light from the Hue Bridge.
This gamut type is added to the state attributes so it shows up in the states panel in the frontend.
And this gamut type is used to set the correct collor conversion method.

starkillerOG committed:
fix snake_case and line to long

starkillerOG committed:
fix inproper placement of line

starkillerOG committed:
docstring

starkillerOG committed:
add gamut to hue_test response

starkillerOG commented on homeassistant/components/light/hue.py:
```diff
@@ -213,10 +215,13 @@ class HueLight(Light):
         self.async_request_bridge_update = request_bridge_update
         self.bridge = bridge
         self.is_group = is_group
+        self.hue_converter = Converter(GamutA)
```
Indeed using the wrong gamut (B instead of A) is what is causing my problems.
I made a few more commits to actually get the gamut type (A, B or C) from the Hue birdge and then use that information to select the right gamut for each bulb.

starkillerOG commented:
@amelchio the problem that this PR fixes is that from whitin home assistant not all colors that some of my philips hue lights can produce, can actually be selected through home assistant.

Currently the home assistant code always assumes color gamut B, however I have some living color lights that are gamut A (as reported by the hue bridge, see [Philips light API](https://developers.meethue.com/develop/hue-api/lights-api/), Sample Response, ["capabilities"]["control"]["colorgamuttype"]). Below the three diffrent gammuts as specified by the Philips hue API (see [Core Concepts, Colors Get More Complicated](https://developers.meethue.com/develop/get-started-2/core-concepts/) section):
![afbeelding](https://user-images.githubusercontent.com/14811189/49786316-a2188a00-fd23-11e8-9b2e-847d72807385.png)


Now if I have a light capable of gamut A, and i select pure blue assuming the light can only produce gamut B, the conversion that home assistant does will tell the light to turn the color specified somewhere at the left bottom red triangle (gamut B). However the light itself is actually only capable of colors wihin the blue triangle, so the hue bridge will correct that xy value as sent by home assistant to something wihin the blue triangle (gamut A) and you will end up somewhere in the more purple section while the light can actually produce a much more blue color (as seen in the blue triangle).

To illustrate this I issues the light.turn_on service within hass with pure blue as attribute for both the current homeassistant code and using this PR.

```
service: light.turn_on
{
  "entity_id": "light.hue_light_id",
  "rgb_color": [0,0,255]
}
```

The resulting color as given in the official Hue android application after issueing the above command using the current HomeAssistant code:
![screenshot_20181211-084443_hue - orig](https://user-images.githubusercontent.com/14811189/49786791-30414000-fd25-11e8-8db0-4ce6bf06e5f5.jpg)
![screenshot_20181211-084735_hue - orig](https://user-images.githubusercontent.com/14811189/49786792-30414000-fd25-11e8-8c6c-ecab6757753c.jpg)

The resulting color as given in the official Hue android application after issueing the above command using the code of this PR (using the correct gamut A for this light):
![screenshot_20181211-084407_hue](https://user-images.githubusercontent.com/14811189/49786850-64b4fc00-fd25-11e8-8c2a-5bfbdde04aa1.jpg)
![screenshot_20181211-084443_hue](https://user-images.githubusercontent.com/14811189/49786848-64b4fc00-fd25-11e8-9157-41803edd85df.jpg)

@amelchio I hope this makes it more clear what I am trying to fix with this PR.

starkillerOG commented:
@amelchio can you help me a bit with this error from travis:
The hue_tests are stuck because light __init__ now needs self.device_info to get the color gamut of the light. within the device_info function the `self.light.raw['capabilities']['control']` is not available in the test enviroment and theirfore an error occurs.

is it possible to replace the self.device_info function with something only returning "fake' information in the test_hue enviroment?
I think it can be done with mock(), but I have never used it and can't seem to figure out how to do it.

The code does work in practice when using it with a real hue bridge, it is only the test that fails.

Mariusthvdb commented:
HI @starkillerOG ,

Don't know how I can help, but fyi, I experience the issue most with the hue color led strips. Even the regular and predefined Hue scenes seem not to work on these and make Savanna Sunset look terribly off and way to bright/yellow..
It does this in the Hue App also, so I am not sure if this is a HA thing or a Hue thing?

Please let me know if you need anything to try and test.

amelchio commented:
That surely was more details 😃

You cannot be right about Gamut B, though, or you would be unable to set green colors at all?

We convert like this:
```text
>>> import homeassistant.util.color as color
>>> color.color_RGB_to_xy(0,0,255)
(0.136, 0.04)
```

... which is well outside any of the triangles. Now, if the Hue hardware is unable to find the closest color from that, I guess it makes sense to correct the value before handing it to the Hue bridge. However, as the gamut triangle is very bulb specific, I wonder if that is a job for `aiohue`.

starkillerOG commented:
@amelchio you could be right, not sure what is going on with the color conversion then.
But I do see that converting in HA results in much more acurate colors as you can see in the pictures above.

The gamut triangle of each bulb can be read for the hue bridge.
It provides simply the gamut type A, B, or C. In the documentation of philips hue it states these 3 are the only gamuts that philips hue uses (screenshot of[ Philips develop doc](https://developers.meethue.com/develop/hue-api/supported-devices/)):
![afbeelding](https://user-images.githubusercontent.com/14811189/49801167-bcfdf500-fd49-11e8-8a62-9fd295ad01b5.png)

However the Hue bridge also provides the cordinates of the color gamut in its response as you can see in the [full response under "capabilities" "control" "colorgamut](https://developers.meethue.com/develop/hue-api/lights-api/)" like:
```
    {
    "1": {
            "state": {
                "on": false,
                "bri": 1,
                "hue": 33761,
                "sat": 254,
                "effect": "none",
                "xy": [
                    0.3171,
                    0.3366
                ],
                "ct": 159,
                "alert": "none",
                "colormode": "xy",
                "mode": "homeautomation",
                "reachable": true
            },
            "swupdate": {
                "state": "noupdates",
                "lastinstall": "2018-01-02T19:24:20"
            },
            "type": "Extended color light",
            "name": "Hue color lamp 7",
            "modelid": "LCT007",
            "manufacturername": "Philips",
            "productname": "Hue color lamp",
            "capabilities": {
                "certified": true,
                "control": {
                    "mindimlevel": 5000,
                    "maxlumen": 600,
                    "colorgamuttype": "B",
                    "colorgamut": [
                        [
                            0.675,
                            0.322
                        ],
                        [
                            0.409,
                            0.518
                        ],
                        [
                            0.167,
                            0.04
                        ]
                    ],
                    "ct": {
                        "min": 153,
                        "max": 500
                    }
                },
                "streaming": {
                    "renderer": true,
                    "proxy": false
                }
            },
            "config": {
                "archetype": "sultanbulb",
                "function": "mixed",
                "direction": "omnidirectional"
            },
            "uniqueid": "00:17:88:01:00:bd:c7:b9-0b",
            "swversion": "5.105.0.21169"
        }
    }
```

Theirfore it is not that bulb specific, their schould only be 3 options. I think it is way easier to just do the color conversion in HASS than to outsource that to aiohue.

@amelchio Do you think this PR could be merged like this if I fix the test_hue?

starkillerOG commented:
@Mariusthvdb As you can see in the pictures below the light strips use color gamut A or C. Theirfore it could verry well be that you are experiancing the same problem as me and that switching to the correct gamut could improve the color acuracy for you.
However I doubt HomeAssistant could do better than the official philips hue app, I am sure philips has done the appropriate conversions in their app so I think the best we can hope for is getting the same experiance of color accuracy in HomeAssistant as you can get in the philips hue app.

![afbeelding](https://user-images.githubusercontent.com/14811189/49801691-2599a180-fd4b-11e8-9d39-ddb5eba322e4.png)
![afbeelding](https://user-images.githubusercontent.com/14811189/49801714-34805400-fd4b-11e8-95db-282ecdbad76a.png)

@Mariusthvdb Would be intresting if you could download the homeassistant/components/light/hue.py from this PR and place it in your custom_compents dir and see if that inproves the color accuracy of your philips hue light strips if you control them from within HomeAssistant (after a restart of HomeAssistant).

amelchio commented:
First, IMO we should not import a library that duplicates functionality we already have; we should extend our own support or migrate our utility function to use the library.

Second, the A/B/C gamuts are very platform specific and I think the right place to put this functionality would be in `aiohue.light.RGB_to_xy()`, `aiohue.light.clamp_xy()` or something like that. But @balloob wrote the rules and he also wrote aiohue so it is probably better that he comments.

starkillerOG commented:
@amelchio alright I see your point.
So if I understand correctly their are 2 options:

1) move the code to determine the Gamut of the light to aiohue, make a aiohue function to convert from RGB to xy using the Gamut of the light (using the rgbxy function just as done in the current PR).

2) Do not move anything to aiohue, but extend the color util of HomeAssistant to be capable of the following:
 - be able to input a color gamut (3 times an xy-point, [[rx,ry],[gx,gy],[bx,by]] )
 - be able to determine if xy color is within color gamut (triangle)
 - be able to calulate the closest point from outside a color gamut triangle to the edge of the color gamut triangle.

I just looked at rgbxy how it is done their, bassically it is just adding mathemetics. It would be about 65 lines of aditional code in the color utility. I don't see how you could do it any more compact than what is done in rgbxy, it is done quite need in rgbxy, so it would basically be copy pasting that code.
Of course you could also import the rgbxy module in the color utility but I think it would then be better to just copy paste the 65 lines into color util.

Which option would you prefer @amelchio and @balloob?

starkillerOG commented:
@balloob Is it alright if I extend the color util? I would like to start working on that code, but I first want your approval.

balloob commented:
aiohue is just a thin wrapper around the Hue API. It should not contain any conversion logic. It would be the right place to store model mapping to color gamuts. 

I am not an expert in color gamuts (and neither want to be 😉), but if it is a problem that we'll experience with other lights, adding it to color util seems appropriate.

starkillerOG commented:
Thank you for your replay.
I will start working on implementing gamuts in the Color util.

starkillerOG committed:
Inplement color gamuts

starkillerOG committed:
Add Gamut capability

starkillerOG committed:
Fix styling

starkillerOG committed:
Fix function discriptions

starkillerOG committed:
fix styling

starkillerOG committed:
fix return float

starkillerOG committed:
fix over-indent

starkillerOG committed:
increase xy conversion precision

starkillerOG committed:
Use new util.color gamut instead of rgbxy

starkillerOG committed:
fix bracket

starkillerOG committed:
remove rgbxy

starkillerOG committed:
remove rgbxy

starkillerOG committed:
remove rgbxy

starkillerOG committed:
Revert higher xy precision

Increasing precision on the xy conversion causes a lot of errors in the test scripts because all conversions will have a digit more.
Theirfore will not do this now.

starkillerOG committed:
include gamut

starkillerOG committed:
fix tests

starkillerOG committed:
fix tests

starkillerOG committed:
fix hue_test

starkillerOG commented:
@balloob and @amelchio, I just finished the rework of this PR.
I reverted the use of the rgbxy module.
Now I have implemented full color gamut capability in the util.color of HomeAssistant.
You can supply an optional tupple of the form ((xR,yR),(xG,yG),(xB,Yb)) to the conversion functions in util.color.
If you do not supply this tupple the conversion will work exactly as it has worked before this PR.
I also adapted the philips hue component to use this new color gamut capability to improve the color accuracy of the philips hue lights.
I also added the gamut information that is normally retriefed from the Hue Hub to the hue_test.py such that all test can pass.

@balloob or @amelchio, can you have another look at this PR?
If everything is fine, can one of you merge the PR?

amelchio commented on homeassistant/components/light/hue.py:
```diff
@@ -310,8 +327,11 @@ def device_info(self):
             # Not yet exposed as properties in aiohue
             'sw_version': self.light.raw['swversion'],
+            'color_gamut_type': light_spec['colorgamuttype'],
+            'color_gamut': light_spec['colorgamut'],
```
The device registry has specific attributes, you should not just add your own stuff. https://developers.home-assistant.io/docs/en/device_registry_index.html

amelchio commented on homeassistant/util/color.py:
```diff
@@ -174,24 +182,26 @@ def color_name_to_rgb(color_name: str) -> Tuple[int, int, int]:
-    G = iG / 255
+    R = iR / 255.0
+    B = iB / 255.0
+    G = iG / 255.0
```
This is not needed, Python 3 has real division by default.

amelchio reviewed (changes requested):
I like this better. I am not sure that the gamut stuff will be used much but it seems to fit well with the color utilities.

If you copied substantial portions from `rgbxy`, I think we have to include the license as well. I am not sure how we usually handle that.

starkillerOG committed:
python 3 already has real division

starkillerOG committed:
tidy up color gamut

starkillerOG committed:
remove unused code

starkillerOG commented:
@amelchio Thanks for your revieuw, I just improved the code to take into acount your comments.
I think I have resolved all issues.
Can you find any more faults, or do you think this code can now be merged?

About the license of rgbxy, I did use rgbxy as a basis for the code in the util.color, However I did modify the code to fit the needs of HomeAssistant. If you look at the rest of the code, the xy conversions were already based on the code provided by philips, wich is also the case for rgbxy, theirfore large portions were already identical (including comment strings):
```
# Converted to Python from Obj-C, original source from:
# http://www.developers.meethue.com/documentation/color-conversions-rgb-xy
```
I will add simular comments as already done in util.color to indicate the parts I havilly based on rgbxy, can you have a look at that and see if that is okay?

amelchio commented:
I will review later.

I did not note the additional math already being present in the Philips examples. I think we are fine about the license then.

starkillerOG committed:
add acknowledgement to rgb

starkillerOG commented:
@amelchio I just added 3 lines of comments with an acknowledgement to rgbxy just to be sure:
```
# The following 5 functions are adapted from rgbxy provided by Benjamin Knight
# License: The MIT License (MIT), 2014.
# https://github.com/benknight/hue-python-rgb-converter
```
I think we are definetly fine if we add these lines, and the util.color already contains simular comments about other persons/code that was adapted, so I think this is the right way to do it.

starkillerOG commented:
@balloob can you merge this PR?
I think everything is fine now with the new color util implementation.
It has been running smooth on my system for a week.

starkillerOG committed:
Add effect property

starkillerOG committed:
fix effect property for groups

starkillerOG commented:
@amelchio can you approve the review?

amelchio reviewed (commented):
I added a few comments but we need @balloob to make the call on this since he is the code owner for Hue.

starkillerOG commented on homeassistant/components/light/hue.py:
```diff
@@ -290,6 +294,11 @@ def supported_features(self):
         return SUPPORT_HUE.get(self.light.type, SUPPORT_HUE_EXTENDED)
 
+    @property
+    def effect(self):
```
Correct, this is unrelated to the color gamut stuff, but I noticed it when I was bussy with the Hue component. So i thought I would just fix that fast.
Is a seprate PR nessesarry for that, or can it just be included in one go?

amelchio replied:
You can keep it now but it's generally best to keep unrelated changes in separate PRs. It makes reviews go faster which makes you happy :)

starkillerOG replied:
Alright, thank you.
sorry for the inconveniance

starkillerOG commented on homeassistant/components/light/hue.py:
```diff
@@ -397,4 +408,25 @@ def device_state_attributes(self):
+        if self.gamut_typ:
+            attributes[ATTR_GAMUT_TYP] = self.gamut_typ
+        if self.gamut:
+            attributes[ATTR_GAMUT] = self.gamut
```
True they can never change, but I wanted a way to lookup the color gamut inside the frontend of homeassistant.
Now you can see them under the dev states panel.

MartinHjelmare replied:
State attributes are for dynamic info related to the state of the entity.

starkillerOG replied:
@MartinHjelmare alright, do you know any other place where I can put this info that is accessible through the front end?
Although things like the `supported_features`, `friendly_name`, `effect_list`, `min_mireds`, `max_mireds` and `is_hue_group` are all vissable in the dev states panel for a Hue light, therefore I think those are also state attributes and those are also not dynamic but just set once when HomeAssistant starts.

MartinHjelmare replied:
There will always be exceptions to a rule. It doesn't mean the rule should be broken or is bad. And rules evolve over time. All parts of the source don't conform to current standards and rules.

`device_info` property would be appropriate in my opinion, but as far as I know, we can't add any key, but need to standardize those. Ask about this in discord chat, and if we need to standardize, it will require an architecture issue discussion and approval.

starkillerOG replied:
@MartinHjelmare I wil just remove this part of the code for now.
It is not essential for opperation, just nice to be able to see what kind of lights you have.

starkillerOG commented on homeassistant/components/light/hue.py:
```diff
@@ -397,4 +408,25 @@ def device_state_attributes(self):
+        try:
+            light_spec = self.light.raw['capabilities']['control']
+            color_gamut_type = light_spec['colorgamuttype']
+            color_gamut = tuple([tuple(x) for x in light_spec['colorgamut']])
```
@balloob sad in a previous post:
> aiohue is just a thin wrapper around the Hue API. It should not contain any conversion logic. It would be the right place to store model mapping to color gamuts.

So I assumed he didn't want it in the aiohue, but i would be perfectly happy moving this function to aiohue.
What do you prefer @balloob?

starkillerOG commented on homeassistant/util/color.py:
```diff
@@ -161,6 +162,13 @@
 
+# Represents the Gamut of a light, ((xR,yR), (xG,yG), (xB,yB))
+GamutType = Tuple[Tuple[float, float], Tuple[float, float],
+                  Tuple[float, float]]
```
Good suggestion, I am already working an adapting the code to use attrs.
Indead that looks much better

starkillerOG commented on homeassistant/util/color.py:
```diff
@@ -161,6 +162,13 @@
+                  Tuple[float, float]]
+
+# Represents a CIE 1931 XY coordinate pair.
+XYPoint = namedtuple('XYPoint', ['x', 'y'])
```
working on it

starkillerOG committed:
use attr for XYPoint and GamutType

starkillerOG committed:
blank lines

starkillerOG committed:
use new attr types

starkillerOG committed:
chnage comments into docstrings

starkillerOG committed:
fix typo

starkillerOG committed:
blanck line after docstring

starkillerOG committed:
Remove Gamut info from state attributes

starkillerOG committed:
Remove gamut attributes

balloob commented:
That one should be added to aiohue as it's helpful for other consumers of the lib too. Tag me in the PR and I will expedite review.

starkillerOG committed:
move self.get_gamut() to aiohue

starkillerOG commented:
I see the hue files have been moved in the latest commit.
homeassistant/components/light/hue.py --> homeassistant/components/hue/light.py
tests/components/light/test_hue.py --> tests/components/hue/test_light.py

After @balloob merged the PR in the aiohue library, I will close this PR and re-open a new PR that uses the new location of the hue files.
(I do not know how to resolve this conflict, so will just close and re-open)

starkillerOG commented:
@balloob I just closed this PR and opend a new PR.
I checked everything and it schould be all fine now.
I also updated the requirements to now use aiohue v1.8.0
